File: examples/example1/example1.cpp

#include "example1.hpp"

namespace example1 {

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

A::~A() noexcept {}
B::~B() noexcept {}


GenA::GenA( A const& arg1 )
  :std::tuple< A >( arg1 ) {}

void GenA::accept( Visitor& visitor ) {
  visitor.visitGenA( *this );
}
void GenA::accept( ConstVisitor& visitor ) const {
  visitor.visitGenA( *this );
}


BToA::BToA( B const& arg1 )
  :std::tuple< B >( arg1 ) {}

void BToA::accept( Visitor& visitor ) {
  visitor.visitBToA( *this );
}
void BToA::accept( ConstVisitor& visitor ) const {
  visitor.visitBToA( *this );
}


GenAB::GenAB( B const& arg1 )
  :std::tuple< B >( arg1 ) {}

void GenAB::accept( Visitor& visitor ) {
  visitor.visitGenAB( *this );
}
void GenAB::accept( ConstVisitor& visitor ) const {
  visitor.visitGenAB( *this );
}


Fin::Fin(  )
  :std::tuple<  >(  ) {}

void Fin::accept( Visitor& visitor ) {
  visitor.visitFin( *this );
}
void Fin::accept( ConstVisitor& visitor ) const {
  visitor.visitFin( *this );
}

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

std::ostream& operator <<( std::ostream &out, A const& self ) {
  class Visitor : public A::ConstVisitor {
  public:
    std::ostream* out_;
    Visitor( std::ostream& out ):out_( &out ){}
    void visitGenA( GenA const& host ) {
      *out_ << host;
    }
    void visitBToA( BToA const& host ) {
      *out_ << host;
    }
  } visitor( out );
  self.accept( visitor );
  return out;
}

std::ostream& operator <<( std::ostream &out, B const& self ) {
  class Visitor : public B::ConstVisitor {
  public:
    std::ostream* out_;
    Visitor( std::ostream& out ):out_( &out ){}
    void visitGenAB( GenAB const& host ) {
      *out_ << host;
    }
    void visitFin( Fin const& host ) {
      *out_ << host;
    }
  } visitor( out );
  self.accept( visitor );
  return out;
}


std::ostream& operator <<( std::ostream& out, GenA const& self ) {
  out << "GenA("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, BToA const& self ) {
  out << "BToA("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, GenAB const& self ) {
  out << "GenAB("<<  std::get< 0 >( self ) << ")";
  return out;
}

std::ostream& operator <<( std::ostream& out, Fin const& self ) {
  out << "Fin(" << ")";
  return out;
}

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

Node1::Node1() {}

Node2::Node2( A const& content_ ) :content( content_ ) {}

Node3::Node3( B const& content_ ) :content( content_ ) {}

Node4::Node4( B const& content_ ) :content( content_ ) {}

Node5::Node5() {}

Node6::Node6( A const& content_ ) :content( content_ ) {}

Node7::Node7() {}

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

std::shared_ptr< State< Node1 > > begin() {
  std::shared_ptr< State<> > bottom( new State<>() );
  return State< Node1 >::make( Node1(), bottom );
}

///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

}

File: examples/example1/Example1.hs

{-# LANGUAGE MultiParamTypeClasses #-}
{-# LANGUAGE FunctionalDependencies #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE UndecidableInstances #-}

module Example1 where

-------------------------------------------------------------------------------

-- grammar definition

-- genA : A -> "a()" A
-- BToA : A -> B
-- genAB : B -> "a()" B "b()"
-- fin : B -> eps

-------------------------------------------------------------------------------

type Program r a = (a -> r) -> r

program :: a -> Program r a
program a = \k -> k a

-------------------------------------------------------------------------------

-- AST nodes

data A
  = GenA A
  | BToA B
  deriving (Show)

data B
  = GenAB B
  | Fin
  deriving (Show)

-------------------------------------------------------------------------------

-- terminal symbols

class ATransition s t | s -> t where
  a :: s -> Program r t

class BTransition s t | s -> t where
  b :: s -> Program r t

-------------------------------------------------------------------------------

-- automaton states

data Node1 prev = Node1 prev

data Node2 prev = Node2 prev A

data Node3 prev = Node3 prev B

data Node4 prev = Node4 prev B

data Node5 prev = Node5 prev

data Node6 prev = Node6 prev A

data Node7 prev = Node7 prev

-------------------------------------------------------------------------------

-- shift transitions

instance {-# OVERLAPS #-} ATransition (Node1 prev) (Node5 (Node1 prev)) where
  a prev = program (Node5 prev)

instance {-# OVERLAPS #-} ATransition (Node5 prev) (Node5 (Node5 prev)) where
  a prev = program (Node5 prev)


instance {-# OVERLAPS #-} BTransition (Node4 prev) (Node7 (Node4 prev)) where
  b prev = program (Node7 prev)

-------------------------------------------------------------------------------

-- reduces

class Reduce s t | s -> t where
  reduce :: s -> t


-- genA : A -> "a()" A

instance Reduce (Node6 (Node5 (Node1 prev))) (Node2 (Node1 prev)) where
  reduce (Node6 (Node5 prev) x1) = (Node2 prev (GenA x1))

instance Reduce (Node6 (Node5 (Node5 prev))) (Node6 (Node5 prev)) where
  reduce (Node6 (Node5 prev) x1) = (Node6 prev (GenA x1))


-- BToA : A -> B

instance Reduce (Node3 (Node1 prev)) (Node2 (Node1 prev)) where
  reduce (Node3 prev x1) = (Node2 prev (BToA x1))

instance Reduce (Node4 (Node5 prev)) (Node6 (Node5 prev)) where
  reduce (Node4 prev x1) = (Node6 prev (BToA x1))


-- genAB : B -> "a()" B "b()"

instance Reduce (Node7 (Node4 (Node5 (Node1 prev)))) (Node3 (Node1 prev)) where
  reduce (Node7 (Node4 (Node5 prev) x1)) = (Node3 prev (GenAB x1))

instance Reduce (Node7 (Node4 (Node5 (Node5 prev)))) (Node4 (Node5 prev)) where
  reduce (Node7 (Node4 (Node5 prev) x1)) = (Node4 prev (GenAB x1))


-- fin : B -> eps

instance Reduce (Node1 prev) (Node3 (Node1 prev)) where
  reduce prev = (Node3 prev (Fin))

instance Reduce (Node5 prev) (Node4 (Node5 prev)) where
  reduce prev = (Node4 prev (Fin))


-- Reduce -> Transition

instance {-# OVERLAPS #-} (Reduce r s, ATransition s t) => ATransition r t where
  a r = a (reduce r)

instance {-# OVERLAPS #-} (Reduce r s, BTransition s t) => BTransition r t where
  b r = b (reduce r)

-------------------------------------------------------------------------------

-- initial state

begin :: Program r (Node1 ())
begin = program (Node1 ())

-------------------------------------------------------------------------------

class End s r | s -> r where
  end :: s -> r

instance {-# OVERLAPS #-} End (Node2 prev) A where
  end (Node2 _ r) = r

instance {-# OVERLAPS #-} (Reduce r s, End s t) => End r t where
  end r = end (reduce r)

-------------------------------------------------------------------------------

